### PR TITLE
Use Rubix async read mode by default

### PR DESCRIPTION
**Pull request by sopel39**

### CODE REVIEW

**Comment by losipiuk** [comment]
Please use correct naming. It used to be "parallel warmup" not "parallel read". But we changed the naming since then "async read mode'

File: presto-hive/src/test/java/io/prestosql/plugin/hive/rubix/TestRubixCaching.java [hunk 1]
```diff
@@ -334,14 +351,22 @@ private Path getStoragePath(String path)
     }
 
     private long getRemoteReadsCount()
-            throws Exception
     {
-        return (long) BEAN_SERVER.getAttribute(new ObjectName("rubix:name=stats,catalog=catalog"), "RemoteReads");
+        try {
```
Comment (losipiuk) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/rubix/TestRubixCaching.java:400]: nit: move to previous commit (ignore if disagree)

File: presto-hive/src/test/java/io/prestosql/plugin/hive/rubix/TestRubixCaching.java [hunk 2]
```diff
@@ -110,9 +115,17 @@ private FileSystem getNonCachingFileSystem()
         return environment.getFileSystem(context, cacheStoragePath);
     }
 
-    private RubixConfigurationInitializer initializeRubix()
+    private void initializerCachingFileSystem(RubixConfig rubixConfig)
```
Comment (losipiuk) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/rubix/TestRubixCaching.java:118]: typo: s/initializer/initialize/

File: presto-hive/src/test/java/io/prestosql/plugin/hive/rubix/TestRubixCaching.java [hunk 3]
```diff
@@ -122,13 +135,16 @@ private RubixConfigurationInitializer initializeRubix()
...
         rubixConfig.setCacheLocation(Joiner.on(",").join(
                 cacheDirectories.stream()
                         .map(java.nio.file.Path::toString)
                         .collect(toImmutableList())));
         RubixConfigurationInitializer rubixConfigInitializer = new RubixConfigurationInitializer(rubixConfig);
-        HdfsConfigurationInitializer configurationInitializer = new HdfsConfigurationInitializer(config, ImmutableSet.of());
+        HdfsConfigurationInitializer configurationInitializer = new HdfsConfigurationInitializer(
+                config,
+                ImmutableSet.of(
+                        // make Rubix to fetch data immediately in parallel mode
```
Comment (losipiuk) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/rubix/TestRubixCaching.java:146]: // fetch data immediately in async mode

File: presto-hive/src/test/java/io/prestosql/plugin/hive/rubix/TestRubixCaching.java [hunk 4]
```diff
@@ -159,6 +175,8 @@ private FileSystem getCachingFileSystem(
                         (dynamicConfig, ignoredContext, ignoredUri) -> {
                             // make sure that dummy cluster manager is used
                             setPrestoClusterManager(dynamicConfig, DummyClusterManager.class.getName());
+                            // ensure new caching FS (with newly initialized BookKeeper) is created for each test case
+                            setCacheKey(dynamicConfig, randomUUID().toString());
```
Comment (losipiuk) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/rubix/TestRubixCaching.java:179]: I am not a big fan that you are overriding the the cache key set by `RubixConfigurationInitializer`. I would rather prefer forcing filesystem cache flushing before each test. Is that possible?

File: presto-hive/src/test/java/io/prestosql/plugin/hive/rubix/TestRubixCaching.java [hunk 5]
```diff
@@ -169,12 +187,30 @@ private FileSystem getCachingFileSystem(
...
             throws IOException
     {
+        nonCachingFileSystem.close();
+    }
+
+    @AfterMethod(alwaysRun = true)
+    public void closeRubix()
+            throws IOException
+    {
+        if (cachingFileSystem == null) {
```
Comment (losipiuk) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/rubix/TestRubixCaching.java:197]: This should be part of lambda passed to `closer.register` below.

File: presto-hive/src/test/java/io/prestosql/plugin/hive/rubix/TestRubixCaching.java [hunk 6]
```diff
@@ -198,18 +234,22 @@ public void testCoordinatorNotJoining()
...
 
-    @Test
-    public void testCacheRead()
+    @Test(dataProvider = "readMode")
+    public void testCacheRead(ReadMode readMode)
             throws Exception
     {
-        Path file = getStoragePath("some_file");
+        initializerCachingFileSystem(new RubixConfig().setReadMode(readMode));
+        byte[] randomData = new byte[(int) SMALL_FILE_SIZE.toBytes()];
```
Comment (losipiuk) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/rubix/TestRubixCaching.java:242]: nit: migrating to randomData is unrelated. Would be nice to have separate commit with that. (not enforcing)

File: presto-hive/src/test/java/io/prestosql/plugin/hive/rubix/TestRubixCaching.java (hunk 5 above)
Comment (sopel39) [reply presto-hive/src/test/java/io/prestosql/plugin/hive/rubix/TestRubixCaching.java:197]: Added explicit `cachingInitialized` field

File: presto-hive/src/test/java/io/prestosql/plugin/hive/rubix/TestRubixCaching.java (hunk 6 above)
Comment (sopel39) [reply presto-hive/src/test/java/io/prestosql/plugin/hive/rubix/TestRubixCaching.java:242]: It's kind of related. `getAsyncDownloadedMb` is the only `BookKeeper` metric that is updated after fetch request is completed. It requires at least 1MB, so I needed to update the test.

File: presto-hive/src/test/java/io/prestosql/plugin/hive/rubix/TestRubixCaching.java [hunk 7]
```diff
@@ -106,9 +114,18 @@ private FileSystem getNonCachingFileSystem()
         return environment.getFileSystem(context, cacheStoragePath);
     }
 
-    private RubixConfigurationInitializer initializeRubix()
+    private void initializeCachingFileSystem(RubixConfig rubixConfig)
             throws IOException
     {
+        cachingFileSystem = getCachingFileSystem(initializeRubix(rubixConfig));
+        cachingInitialized = true;
```
Comment (losipiuk) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/rubix/TestRubixCaching.java:121]: It does not work correctly. Note that you can get exception from `getCachingFileSystem`, after `initializeRubix` is already called. 
Then the `closeRubix` will not be called and cleanup will not be made. Not sure if that is big of a deal. But if we want the cleanup to happen we need to guard on individual fields which are closed/cleaned up, not on whole routine.

File: presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveCaching.java [hunk 8]
```diff
@@ -82,6 +83,25 @@ private void testReadFromTable(String tableNameSuffix)
...
+        StringBuilder randomDataBuilder = new StringBuilder();
+        Random random = new Random();
+        for (int i = 0; i < 500_000; ++i) {
+            randomDataBuilder.append(random.nextInt(10));
+        }
+        String randomData = randomDataBuilder.toString();
+
+        query("DROP TABLE IF EXISTS " + tableName);
+        // use `format` to overcome SQL query length limit
+        query("CREATE TABLE " + tableName + " WITH (format='TEXTFILE') AS SELECT format('%1$s%1$s%1$s%1$s%1$s', '" + randomData + "') as col");
```
Comment (losipiuk) [inline presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveCaching.java:103]: nit: you can generate format using `5` so it is obvious that returned String matches what is inserted into table.

File: presto-hive/src/test/java/io/prestosql/plugin/hive/rubix/TestRubixCaching.java (hunk 7 above)
Comment (sopel39) [reply presto-hive/src/test/java/io/prestosql/plugin/hive/rubix/TestRubixCaching.java:121]: You are right. I will add explicit null checks
